packages/fxa-shared/test/experiments/base.js

```diff
@@ -0,0 +1,258 @@
+        `${trueCount} is too far from ${fiftyPercent}`
+      );
+    });
+  });
+
+  describe('uniformChoice', () => {
+    it('distributes members uniformly amongst grouping-rules', () => {
+      const counts = {
+        control: 0,
+        treatment: 0,
+      };
+
+      for (var i = 0; i < ITERATIONS; ++i) {
+        const choice = experiment.uniformChoice(
+          ['control', 'treatment'],
+          Math.random()
+        );
+        counts[choice]++;
+      }
+
+      const leeway = Math.floor(ITERATIONS * (LEEWAY_PERCENTAGE / 100));
+      const fiftyPercent = Math.round(ITERATIONS / 2);
+      const min = fiftyPercent - leeway;
+      const max = fiftyPercent + leeway;
+      assert.ok(
+        min <= counts.control && counts.control <= max,
+        `${counts.control} is too far from ${fiftyPercent}`
+      );
+      assert.ok(
+        min <= counts.treatment && counts.treatment <= max,
+        `${counts.treatment} is too far from ${fiftyPercent}`
+      );
+
+      assert.equal(counts.control + counts.treatment, ITERATIONS);
+    });
+  });
+
+  describe('choose', () => {
+    it('must be overridden', () => {
+      assert.throws(function() {
+        experiment.choose();
+      }, 'choose must be overridden');
+    });
+
+    it('throws if deprecated', () => {
+      assert.throws(() => {
+        experiment.name = 'oldExperiment';
+        experiment.deprecated = true;
+        experiment.choose();
+      }, 'Experiment deprecated: oldExperiment');
+    });
+  });
+
+  describe('isTestEmail', () => {
+    [
+      'tester@mozilla.com',
+      'testuser@mozilla.org',
+      'tester@softvision.ro',
+      'tester@softvision.com',
+    ].forEach(email => {
+      it(`returns 'true' for test email: ${email}`, () => {
+        assert.isTrue(experiment.isTestEmail(email));
+      });
+    });
+
+    ['tester@google.com', 'tester@mozilla.es'].forEach(email => {
+      it(`returns false for other non-test email: ${email}`, () => {
+        assert.isFalse(experiment.isTestEmail(email));
+      });
+    });
+  });
+
+  describe('one experiment choose another', () => {
+    /**
+     * See #5378. This test is to ensure the hashing function has a uniform distribution
+     * when one experiment chooses another. We originally chose crc32 as the hashing
+     * function. This worked fine when experiment.choose for two experiments were
+     * called independently, but when one experiment was used to choose another,
+     * *all* users of the chosen experiment were placed into the same bucket. md5
+     * doesn't suffer from this problem.
+     */
+    class Experiment1 extends BaseExperiment {
+      constructor() {
+        super();
+        this.name = 'experiment1';
+      }
+
+      choose(subject) {
+        if (subject.experimentChooser.choose(subject) !== this.name) {
+          return false;
+        }
+
+        const GROUPS = ['control', 'treatment'];
+        return this.uniformChoice(GROUPS, subject.uuid);
+      }
```

Comment:
nit: from looking at this impl, `choose` either returns false or a string?  Could add that the method declaration in the base class.

Reply:
TIL that typescript has [union types](https://www.typescriptlang.org/docs/handbook/advanced-types.html#union-types). Allows function to return boolean or string.